src/common/tracing/src/remote_log.rs

```diff
@@ -0,0 +1,301 @@
+                    RemoteLog::flush(flush_buffer, stage_name).await;
+                }
+            }
+        }
+        // receiver close will trigger interval_flush exit
+        receiver.close()
+    }
+
+    pub async fn flush(flush_buffer: Vec<RemoteLogElement>, stage_name: &str) {
+        if flush_buffer.is_empty() {
+            return;
+        }
+        let op = GlobalLogger::instance().get_operator().await;
+        if op.is_none() {
+            return;
+        }
+        let path = format!(
+            "stage/internal/{}/{}.parquet",
+            stage_name,
+            uuid::Uuid::new_v4()
+        );
+        spawn(async move {
+            if let Err(e) = Self::do_flush(op.unwrap(), flush_buffer, &path).await {
+                eprintln!("Failed to flush logs: {:?}", e);
+            }
+        });
+    }
+
+    pub async fn do_flush(
+        op: Operator,
+        flush_buffer: Vec<RemoteLogElement>,
+        path: &str,
+    ) -> Result<()> {
+        let props = WriterProperties::builder()
+            .set_compression(Compression::ZSTD(ZstdLevel::try_new(1)?))
+            .set_dictionary_enabled(false)
+            .set_statistics_enabled(EnabledStatistics::None)
+            .set_bloom_filter_enabled(false)
+            .build();
+        let batch = convert_to_batch(flush_buffer)?;
+        let mut buf = vec![];
+        let mut writer = ArrowWriter::try_new(&mut buf, batch.schema(), Some(props))?;
+        writer.write(&batch)?;
+        writer.close()?;
+        op.write(path, buf).await?;
+        Ok(())
+    }
+
+    fn prepare_log_element(&self, record: &Record) -> RemoteLogElement {
+        let query_id = ThreadTracker::query_id();
+        let mut fields = Map::new();
+        fields.insert("message".to_string(), format!("{}", record.args()).into());
+        for (k, v) in collect_kvs(record.key_values()) {
+            fields.insert(k, v.into());
+        }
+        let messages = serde_json::to_string(&fields).unwrap_or_default();
+        let log_level = record.level().to_string();
+        let timestamp = chrono::Local::now().timestamp_micros();
+        let path = format!(
+            "{}: {}:{}",
+            record.module_path().unwrap_or(""),
+            Path::new(record.file().unwrap_or_default())
+                .file_name()
+                .and_then(|name| name.to_str())
+                .unwrap_or_default(),
+            record.line().unwrap_or(0)
+        );
+
+        RemoteLogElement {
+            timestamp,
+            path,
+            cluster_id: self.cluster_id.clone(),
+            node_id: self.node_id.clone(),
+            query_id: query_id.cloned(),
+            log_level,
+            messages,
+        }
+    }
+}
+
+impl Append for RemoteLog {
+    fn append(&self, record: &Record) -> anyhow::Result<()> {
+        let log_element = self.prepare_log_element(record);
+        let _ = self.sender.try_send(Message::LogElement(log_element));
```

Comment:
Does this block and wait until the worker finishes processing the previous message?

Reply:
Yes. This should block and wait. `work` method will alwasy spawn new task, so the block time will not be very short.